src/box/alter.cc

```diff
@@ -1773,7 +1797,14 @@ alter_space_move_indexes(struct alter_space *alter, uint32_t begin,
 					old_def->name, strlen(old_def->name),
 					old_def->type, &old_def->opts,
 					old_def->key_def, alter->pk_def);
-		index_def_update_optionality(new_def, min_field_count);
+		/*
+		 * Optionality is "extended" for the same reason as for indexes
+		 * not depending on primary key.
+		 */
+		if (alter->is_space_upgrade)
+			index_def_extend_optionality(new_def, min_field_count);
+		else
+			index_def_update_optionality(new_def, min_field_count);
```

Comment:
I think there's a much easier way to achieve the same result:

```diff
diff --git a/src/box/alter.cc b/src/box/alter.cc
index 660cd1ef1c90..3753fc7bdcea 100644
--- a/src/box/alter.cc
+++ b/src/box/alter.cc
@@ -2355,6 +2355,11 @@ on_replace_dd_space(struct trigger * /* trigger */, void *event)
 						     old_space->index_count,
 						     def->fields,
 						     def->field_count);
+		if (def->opts.upgrade_def != NULL) {
+			alter->new_min_field_count = MIN(
+				alter->new_min_field_count,
+				old_space->format->min_field_count);
+		}
 		try {
 			(void) new CheckSpaceFormat(alter);
 			(void) new ModifySpace(alter, def);
```

This will extend index optionality on space upgrade. Note that patching the EE repository isn't required because this code is also executed when space upgrade is completed.

Reply:
Thanks, I stick with your solution now.